pkg/cli/sync_object_store.go

```diff
@@ -0,0 +1,115 @@
+	var srcAccessKeySecret string
+
+	var dstHost string
+	var dstAccessKeyID string
+	var dstAccessKeySecret string
+
+	syncObjectStoreCmd := &cobra.Command{
+		Use:   "sync-object-store",
+		Short: "Copies buckets and objects from one object store to another",
+		Run: func(cmd *cobra.Command, args []string) {
+			src, err := minio.New(
+				srcHost,
+				srcAccessKeyID,
+				srcAccessKeySecret,
+				false,
+			)
+			if err != nil {
+				log.Panic(err)
+			}
+
+			dst, err := minio.New(
+				dstHost,
+				dstAccessKeyID,
+				dstAccessKeySecret,
+				false,
+			)
+			if err != nil {
+				log.Panic(err)
+			}
+
+			ctx := context.Background()
+
+			srcBuckets, err := src.ListBuckets()
+			if err != nil {
+				log.Fatalf("Failed to list buckets in %s: %v", srcHost, err)
+			}
+
+			for _, srcBucket := range srcBuckets {
+				fmt.Printf("Syncing %s from %s to %s\n", srcBucket.Name, srcHost, dstHost)
+
+				count, err := SyncBucket(ctx, src, dst, srcBucket.Name)
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				fmt.Printf("Successfully synced %d objects in bucket %s from %s to %s\n", count, srcBucket.Name, srcHost, dstHost)
+			}
+
+			fmt.Printf("Successfully synced %d buckets from %s to %s\n", len(srcBuckets), srcHost, dstHost)
+		},
+	}
+
+	syncObjectStoreCmd.Flags().StringVar(&srcHost, "source_host", "", "Hostname of the source object store")
+	syncObjectStoreCmd.Flags().StringVar(&srcAccessKeyID, "source_access_key_id", "", "Access key ID for the source object store")
+	syncObjectStoreCmd.Flags().StringVar(&srcAccessKeySecret, "source_access_key_secret", "", "Access key secret for the source object store")
+
+	syncObjectStoreCmd.Flags().StringVar(&dstHost, "dest_host", "", "Hostname of the destination object store")
+	syncObjectStoreCmd.Flags().StringVar(&dstAccessKeyID, "dest_access_key_id", "", "Access key ID for the destination object store")
+	syncObjectStoreCmd.Flags().StringVar(&dstAccessKeySecret, "dest_access_key_secret", "", "Access key secret for the destination object store")
+
+	return syncObjectStoreCmd
+}
+
+func SyncBucket(ctx context.Context, src *minio.Client, dst *minio.Client, bucket string) (int, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	count := 0
+
+	exists, err := dst.BucketExists(bucket)
+	if err != nil {
+		return count, fmt.Errorf("Failed to check if bucket %q exists in destination: %v", bucket, err)
+	}
+	if !exists {
+		if err := dst.MakeBucket(bucket, ""); err != nil {
+			return count, fmt.Errorf("Failed to make bucket %q in destination: %v", bucket, err)
+		}
+	}
+
+	srcObjectInfoChan := src.ListObjects(bucket, "", true, ctx.Done())
+
+	for srcObjectInfo := range srcObjectInfoChan {
+		srcObject, err := src.GetObject(bucket, srcObjectInfo.Key, minio.GetObjectOptions{})
+		if err != nil {
+			return count, fmt.Errorf("Get %s from source: %v", srcObjectInfo.Key, err)
+		}
+
+		_, err = dst.PutObject(bucket, srcObjectInfo.Key, srcObject, srcObjectInfo.Size, minio.PutObjectOptions{
```

Comment:
what happens if the object already exists in the destination? Should we delete it and recreate it?

Or compare hashes to speed up things if you cancelled and reran the process?

Reply:
This will overwrite whatever object is there. Yes, I can optimize with etag.

Reply:
Actually RGW and Minio ETags don't match for the same object so I don't think this can be optimized.